### ISSUE TITLE
[TextField] Flutter web - Textfield hardware_keyboard error

### ISSUE THREAD

**Issue by Patrick386** [flutter/flutter#126247]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

I am developing Flutter Web and getting a 'hardware_keyboard' error when I continuously input Korean characters in a TextField. This is a very old issue. Is there a way to fix this problem?

### Expected results

When typing Korean characters in a TextField, the expected result is that the characters are inputted correctly and without any errors or crashes. The TextField should function as intended and the app should respond accordingly to the input.

### Actual results

The 'hardware_keyboard' error occurs when typing Korean characters in a TextField, especially when typing continuously or rapidly. This error can cause the app to crash or behave unexpectedly.

### Code sample

<details><summary>Code sample</summary>

```dart
[   return Scaffold(
      body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(
                width: 300.0,
                child: TextField(),
              )
            ],
          )
      ),

    );]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Performing hot restart...
Waiting for connection from debug service on Chrome...
Restarted application in 1,766ms.
Error: Assertion failed: file:///C:/Dev/flutter/packages/flutter/lib/src/services/hardware_keyboard.dart:435:16
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 266:49      throw_
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3        assertFailed
packages/flutter/src/services/hardware_keyboard.dart 435:56                                                                    <fn>
packages/flutter/src/services/hardware_keyboard.dart 442:14                                                                    [_assertEventIsRegular]
packages/flutter/src/services/hardware_keyboard.dart 535:5                                                                     handleKeyEvent
packages/flutter/src/services/hardware_keyboard.dart 881:29                                                                    handleKeyData
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/platform_dispatcher.dart 272:33                                <fn>
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/platform_dispatcher.dart 1168:13                               invoke
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/platform_dispatcher.dart 271:7                                 invokeOnKeyData
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 161:39                                   [_onKeyData]
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 579:36                                   <fn>
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 454:11                                   [_handleEvent]
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 582:7                                    handleEvent
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 97:25                                    <fn>
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 367:37  _checkAndCall
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 372:39  dcall
C:/b/s/w/ir/cache/builder/src/out/host_debug/lib/_engine/engine/keyboard_binding.dart 139:23                                   loggedHandler
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 367:37  _checkAndCall
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 372:39  dcall]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[[√] Flutter (Channel stable, 3.7.12, on Microsoft Windows [Version 10.0.19045.2846], locale ko-KR)
    • Flutter version 3.7.12 on channel stable at C:\Dev\flutter                                  
    • Upstream repository https://github.com/flutter/flutter.git                                  
    • Framework revision 4d9e56e694 (3 weeks ago), 2023-04-17 21:47:46 -0400                      
    • Engine revision 1a65d409c7                                                                  
    • Dart version 2.19.6                                                                         
    • DevTools version 2.20.1

[X] Windows Version (Unable to confirm if installed Windows version is 10 or greater)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\SintPatrick\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.4.5)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.4.33403.182
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] IntelliJ IDEA Ultimate Edition (version 2022.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2022.2.2
    • Flutter plugin version 73.0.2
    • Dart plugin version 222.4167.21

[√] VS Code (version 1.75.0)
    • VS Code at C:\Users\SintPatrick\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.58.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2846]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 112.0.5615.139
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 113.0.1774.35

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
]
```

</details>

**Comment by huycozy** [comment]
Hi @Patrick386 
I try typing text quickly but can not replicate the issue. 

<details>
<summary>Demo - Samsung Galaxy Tab A7 Lite, Android 13</summary>

https://user-images.githubusercontent.com/104349824/236780783-7bf5bcfa-242f-4b80-9a14-27003b87da93.mov

</details>

Can you please share the complete output of `flutter run -v` so that we may verify the issue?

**Comment by Patrick386** [comment]
> 
Hi, 
Please test it on a PC.

**Comment by huycozy** [comment]
Thank you! I can reproduce this issue on Web. The stack trace on stable `3.7.12` is the same as above. But when checking this on the latest master, it's more informative as below logs.

### Additional info

- The issue doesn't happen on Windows or macOS app. 
- It seems only to happen on Web. On Web (macOS) I only see the issue one time and it is not reproducible anymore. Meanwhile, it is reproducible all the time on the Web (Windows) with the steps below

### Steps to reproduce
1. Run the app with single `TextField` widget
2. Enter Korean text to TextField
3. Enter one of these keys: `Space`/`Window`/`Control` (not alphabet key)
4. Check the output log

<details>
<summary>Logs (master channel)</summary>

```console
Error: Assertion failed: file:///C:/WIP/flutter_master/packages/flutter/lib/src/services/hardware_keyboard.dart:474:16
_pressedKeys[event.physicalKey] == event.logicalKey
"A KeyUpEvent is dispatched, but the state shows that the physical key is pressed on a different logical key. If this occurs in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the event: KeyUpEvent#7568c(physicalKey: PhysicalKeyboardKey#7002c(usbHidUsage: \"0x0007002c\", debugName: \"Space\"), logicalKey: LogicalKeyboardKey#00020(keyId: \"0x00000020\", keyLabel: \" \", debugName: \"Space\"), character: null, timeStamp: 0:00:39.648100, synthesized) and the recorded logical key LogicalKeyboardKey#602a3(keyId: \"0x10000070f\", keyLabel: \"Process\", debugName: \"Process\")"
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 288:49      throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3        assertFailed
packages/flutter/src/services/hardware_keyboard.dart 474:56                       <fn>
packages/flutter/src/services/hardware_keyboard.dart 481:14                       [_assertEventIsRegular]
packages/flutter/src/services/hardware_keyboard.dart 601:5                        handleKeyEvent
packages/flutter/src/services/hardware_keyboard.dart 948:29                       handleKeyData
lib/_engine/engine/platform_dispatcher.dart 301:33                                <fn>
lib/_engine/engine/platform_dispatcher.dart 1245:13                               invoke
lib/_engine/engine/platform_dispatcher.dart 300:7                                 invokeOnKeyData
lib/_engine/engine/keyboard_binding.dart 161:39                                   [_onKeyData]
lib/_engine/engine/keyboard_binding.dart 579:36                                   <fn>
lib/_engine/engine/keyboard_binding.dart 457:11                                   [_handleEvent]
lib/_engine/engine/keyboard_binding.dart 582:7                                    handleEvent
lib/_engine/engine/keyboard_binding.dart 98:18                                    <fn>
lib/_engine/engine/keyboard_binding.dart 140:16                                   loggedHandler
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 504:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 509:39  dcall
```
</details>

Labeling the issue for more investigation!

<details>
<summary>flutter doctor -v (stable & master)</summary>

```bash
[√] Flutter (Channel stable, 3.7.12, on Microsoft Windows [Version 10.0.19045.2846], locale en-US)
    • Flutter version 3.7.12 on channel stable at C:\WIP\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4d9e56e694 (6 days ago), 2023-04-17 21:47:46 -0400
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\AndroidSDK
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = C:\AndroidSDK
    • Java binary at: C:\Program Files\Android\Android Studio Electric Eel 2022.1.1\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.4.4)
    • Visual Studio at D:\DOWNLOADWORK\VS2022
    • Visual Studio Community 2022 version 17.4.33213.308
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] Android Studio (version 2022.1)
    • Android Studio at C:\Program Files\Android\Android Studio Electric Eel 2022.1.1
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[√] VS Code (version 1.77.3)
    • VS Code at C:\Users\ADMIN\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.62.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2846]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 112.0.5615.138
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 111.0.1661.54

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.11.0-2.0.pre.73, on Microsoft Windows [Version 10.0.19045.2846],
    locale en-US)
    • Flutter version 3.11.0-2.0.pre.73 on channel master at C:\WIP\flutter_master
    ! Warning: `flutter` on your path resolves to C:\WIP\flutter\bin\flutter, which is not inside
      your current Flutter SDK checkout at C:\WIP\flutter_master. Consider adding
      C:\WIP\flutter_master\bin to the front of your path.
    ! Warning: `dart` on your path resolves to C:\WIP\flutter\bin\dart, which is not inside your
      current Flutter SDK checkout at C:\WIP\flutter_master. Consider adding
      C:\WIP\flutter_master\bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 8c5a1ea728 (9 hours ago), 2023-05-08 14:00:20 -0700
    • Engine revision 8d3a8162b3
    • Dart version 3.1.0 (build 3.1.0-86.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to
      use "git" directly to perform update checks and upgrades.

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at C:\AndroidSDK
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = C:\AndroidSDK
    • Java binary at: C:\Program Files\Android\Android Studio Electric Eel 2022.1.1\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.4.4)
    • Visual Studio at D:\DOWNLOADWORK\VS2022
    • Visual Studio Community 2022 version 17.4.33213.308
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] Android Studio (version 2022.1)
    • Android Studio at C:\Program Files\Android\Android Studio Electric Eel 2022.1.1
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[√] VS Code (version 1.78.0)
    • VS Code at C:\Users\ADMIN\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.62.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2846]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 112.0.5615.138
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 113.0.1774.35

[√] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by htoor3** [comment]
This might not be IME specific as the `hardware_keyboard` error is also occurring in the issue here https://github.com/flutter/flutter/issues/100455 during keyboard combinations.  

I can consistently reproduce this error in MacOS on Firefox when I hit Tab during IME composition for Korean which results in the same mismatch between the physical key (`Tab`) and the logical key (`Process`).  

This is the assertion that fails on keyup events: https://github.com/flutter/flutter/blob/5b2f658ff5ac3179bddc814f95ad30c1bf950e2c/packages/flutter/lib/src/services/hardware_keyboard.dart#L474

@dkwingsmt it's possible that an extra `Process` key is registered when you hit `Tab` or a special character during IME composition and that may be the root of the physical/logical key mismatch.

**Comment by MastProgs** [comment]
I'm having the same problem as this one. this problem seems to be occurring only on the web base.
Windows Desktop, Chrome Web, Edge Web, and all three cases have been tested, and it's always only happening on the Web.
Basically, after entering Korean, if you enter spaces, numbers, or other English words, it suddenly stops and you get the same error as the author said.

```
Exception has occurred.
Error: Assertion failed: file:///F:/Devlop/flutter/packages/flutter/lib/src/services/hardware_keyboard.dart:474:16
_pressedKeys[event.physicalKey] == event.logicalKey
"A KeyUpEvent is dispatched, but the state shows that the physical key is pressed on a different logical key. If this occurs in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the event: KeyUpEvent#20a31(physicalKey: PhysicalKeyboardKey#7002c(usbHidUsage: \"0x0007002c\", debugName: \"Space\"), logicalKey: LogicalKeyboardKey#00020(keyId: \"0x00000020\", keyLabel: \" \", debugName: \"Space\"), character: null, timeStamp: 0:00:25.661099, synthesized) and the recorded logical key LogicalKeyboardKey#602a3(keyId: \"0x10000070f\", keyLabel: \"Process\", debugName: \"Process\")"
```

<details><summary>Error Code Position</summary>

```dart
  void _assertEventIsRegular(KeyEvent event) {
    assert(() {
      const String common = 'If this occurs in real application, please report this '
        'bug to Flutter. If this occurs in unit tests, please ensure that '
        "simulated events follow Flutter's event model as documented in "
        '`HardwareKeyboard`. This was the event: ';
      if (event is KeyDownEvent) {
        assert(!_pressedKeys.containsKey(event.physicalKey),
          'A ${event.runtimeType} is dispatched, but the state shows that the physical '
          'key is already pressed. $common$event');
      } else if (event is KeyRepeatEvent || event is KeyUpEvent) {
        assert(_pressedKeys.containsKey(event.physicalKey),
          'A ${event.runtimeType} is dispatched, but the state shows that the physical '
          'key is not pressed. $common$event');
        assert(_pressedKeys[event.physicalKey] == event.logicalKey, // Here <----
```
</details>


![001](https://github.com/flutter/flutter/assets/14367684/39f89d90-2d93-4546-aeab-6551a361c322)

Based on the above picture, it occurs when other characters such as spaces or numbers are entered at the time when the cursor is still under the '트' after entering the '비트'.


The reason why you make a different example and you don't have anything to do, it just happens in the basic code.

My case is like this

<details><summary>Error Code Position</summary>

```dart
 // just make controller
  final TextEditingController _idController = TextEditingController();
 // and use controller in textfield.
  TextField( controller: _idController,
 // this is all..
```
</details>

There is nothing the user can do because it occurs within the low level of the input, even before the error occurs in the code woven by the direct programmer.

There is no problem in the process of entering only Korean, but 100% of the cases where non-Korean characters such as spacing or numeric input occur.
However, if you reposition your mouse so that the cursor is out of focus while entering Korean, there will be no related problem if you type English, numbers, or special characters.

I think when you type in Korean, the last letter is underlined and the cursor and underline are focused, and at this point, in the case of Korean input, consonant and vowel combinations can occur in various forms.
For example, if you type 'ㄹ' even in the '가' state, you can type '갈', but the letter input may or may not be finished. When you type 'ㄱ' in the '갈' state, it becomes '갉' and it is a part where you wonder if the combination and ending of Korean letters are not done properly.
So, to summarize, if you type in a space bar, shift, English, number, etc. in the process of trying to combine the last focused Korean characters, you can guess that there will be a crash to create characters that you can't really combine.

But I'm not sure why that part only occurs in pc input situations, when I run it on the web. I don't use a mobile emulator, so I can't test other environments.

This problem has been around for quite a while, and I hope it is resolved as soon as possible.

**Comment by htoor3** [comment]
@Patrick386 a fix was pushed by @bleroux to `master` for a similar issue in french keyboards.

Can you verify that this also fixes your issue?
